pallets/moonbeam-orbiters/src/lib.rs

```diff
@@ -0,0 +1,504 @@
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The collator is already added in orbiters program.
+		CollatorAlreadyAdded,
+		/// This collator is not in orbiters program.
+		CollatorNotFound,
+		/// There are already too many orbiters associated with this collator.
+		CollatorPoolTooLarge,
+		/// There are more collator pools than the number specified in the parameter.
+		CollatorsPoolCountTooLow,
+		/// The minimum deposit required to register as an orbiter has not yet been included in the
+		/// onchain storage
+		MinOrbiterDepositNotSet,
+		/// This orbiter is already associated with this collator.
+		OrbiterAlreadyInPool,
+		/// Orbiter cant leave this round
+		OrbiterCantLeaveThisRound,
+		/// This orbiter has not made a deposit
+		OrbiterDepositNotFound,
+		/// This orbiter is not found
+		OrbiterNotFound,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Paid the orbiter account the balance as liquid rewards.
+		OrbiterRewarded {
+			account: T::AccountId,
+			rewards: BalanceOf<T>,
+		},
+		/*SelectedOrbiters {
+			orbiters:
+		}*/
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add an orbiter in a collator pool
+		#[pallet::weight(500_000_000)]
+		pub fn collator_add_orbiter(
+			origin: OriginFor<T>,
+			orbiter: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let collator = ensure_signed(origin)?;
+			let orbiter = T::Lookup::lookup(orbiter)?;
+
+			let mut collator_pool =
+				CollatorsPool::<T>::get(&collator).ok_or(Error::<T>::CollatorNotFound)?;
+			let orbiters = collator_pool.get_orbiters();
+			ensure!(
+				(orbiters.len() as u32) < T::MaxPoolSize::get(),
+				Error::<T>::CollatorPoolTooLarge
+			);
+			if orbiters.iter().any(|orbiter_| orbiter_ == &orbiter) {
+				return Err(Error::<T>::OrbiterAlreadyInPool.into());
+			}
+
+			// Make sure the orbiter has made a deposit. It can be an old orbiter whose deposit
+			// is lower than the current minimum (if the minimum was lower in the past), so we just
+			// have to check that a deposit exists (which means checking that the deposit amount
+			// is not zero).
+			let orbiter_deposit =
+				T::Currency::reserved_balance_named(&T::OrbiterReserveIdentifier::get(), &orbiter);
+			ensure!(
+				orbiter_deposit > BalanceOf::<T>::zero(),
+				Error::<T>::OrbiterDepositNotFound
+			);
+
+			collator_pool.add_orbiter(orbiter);
+			CollatorsPool::<T>::insert(collator, collator_pool);
+
+			Ok(())
+		}
+
+		/// Remove an orbiter from the caller collator pool
+		#[pallet::weight(500_000_000)]
+		pub fn collator_remove_orbiter(
+			origin: OriginFor<T>,
+			orbiter: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let collator = ensure_signed(origin)?;
+			let orbiter = T::Lookup::lookup(orbiter)?;
+
+			let mut collator_pool =
+				CollatorsPool::<T>::get(&collator).ok_or(Error::<T>::CollatorNotFound)?;
+
+			ensure!(
+				collator_pool.remove_orbiter(&orbiter),
```

Comment:
Is it a problem is the orbiter is selected for the current round ? Will it still be supposed to produce block  ?

Reply:
Currently, the orbiter must continue to produce blocks until the next round.
If this is not appropriate, I see 3 possible solutions:
1. Forbidding to remove the orbiter from the pool as long as it is currently selected (allow in the special case where it is the only orbiter left in the pool).
2. Rotate immediately to the next orbiter, but the rewards of the round remain in full for the old orbiter
3. Rotate immediately to the next orbiter, but the rewards of the round go entirely to the new orbiter

Reply:
We can't force an orbiter to produce blocks (or punish them for not doing it) in any case, so this doesn't seem very important to me. In other words, if we do anything to punish them for leaving while "on duty" here, they can simply stop producing blocks anyway and not interact with the orbiters pallet.

I think the best we can do here is give them a pain-free way to let someone else take their spot.